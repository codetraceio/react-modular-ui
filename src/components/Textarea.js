import React from 'react';
import AbstractComponent from './AbstractComponent';

export default class Textarea extends AbstractComponent {
  getModifierObject() {
    return {
      size: this.props.size,
      view: this.props.view,
      color: this.props.color,
      disabled: this.props.disabled
    };
  }

  getStyle() {
    const style = {};
    if (this.props.height) {
      style.height = `${this.props.height}px`;
    }
    return style;
  }

  delegateEvent(eventName) {
    return (event) => {
      if (typeof this.props[eventName] === 'function') {
        this.props[eventName](event.target.value, event);
      }
    };
  }

  renderLabel() {
    return this.props.label ? (
      <div className={this.getElementName('textarea', 'label')}>{this.props.label}</div>
    ) : null;
  }

  render() {
    return (
      <div className={this.getBlockName('textarea', this.getModifierObject())}>
        {this.renderLabel()}
        <textarea
          name={this.props.name}
          placeholder={this.props.placeholder}
          value={this.props.value}
          disabled={this.props.disabled}
          style={this.getStyle()}
          onChange={this.delegateEvent('onChange')}
          onKeyDown={this.delegateEvent('onKeyDown')}
          onKeyUp={this.delegateEvent('onKeyUp')}
          onKeyPress={this.delegateEvent('onKeyPress')}
          onFocus={this.delegateEvent('onFocus')}
          onBlur={this.delegateEvent('onBlur')}
        />
      </div>
    )
  }
}
