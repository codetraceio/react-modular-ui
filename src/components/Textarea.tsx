import React, { useContext, useMemo } from "react";

import { className } from "../utils/className";
import { useEvent } from "../hooks/useEvent";
import { ThemeContext } from "./ThemeContext";

export interface TextareaProps {
  size?: string | number;
  variant?: string;
  color?: string;
  disabled?: boolean;
  height?: string | number;
  label?: string;
  name?: string;
  placeholder?: string;
  value?: string | number;
  theme?: string;
  data?: Record<string, string>;

  onChange?: (
    value: string,
    event: React.ChangeEvent<HTMLTextAreaElement>,
  ) => void;
  onKeyDown?: (
    value: string,
    event: React.KeyboardEvent<HTMLTextAreaElement>,
  ) => void;
  onKeyUp?: (
    value: string,
    event: React.KeyboardEvent<HTMLTextAreaElement>,
  ) => void;
  onFocus?: (
    value: string,
    event: React.FocusEvent<HTMLTextAreaElement>,
  ) => void;
  onBlur?: (
    value: string,
    event: React.FocusEvent<HTMLTextAreaElement>,
  ) => void;
}

export default function Textarea({
  size,
  variant,
  color,
  disabled,
  height,
  label,
  name,
  placeholder,
  value,
  theme,
  data = {},
  onChange,
  onKeyDown,
  onKeyUp,
  onFocus,
  onBlur,
  ...props
}: TextareaProps) {
  const themeContext = useContext(ThemeContext);

  const labelElement = useMemo(() => {
    return label ? (
      <div className={className("textarea", "label")}>{label}</div>
    ) : null;
  }, [label]);

  return (
    <div
      className={className("textarea")}
      data-size={size}
      data-variant={variant}
      data-color={color}
      data-theme={theme ?? themeContext}
      aria-disabled={disabled}
      {...props}
    >
      {labelElement}
      <textarea
        name={name}
        placeholder={placeholder}
        value={value}
        disabled={disabled}
        style={{
          height: height,
        }}
        onChange={useEvent(onChange, data)}
        onKeyDown={useEvent(onKeyDown, data)}
        onKeyUp={useEvent(onKeyUp, data)}
        onFocus={useEvent(onFocus, data)}
        onBlur={useEvent(onBlur, data)}
      />
    </div>
  );
}
