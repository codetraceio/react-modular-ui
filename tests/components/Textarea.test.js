import React from 'react';
import renderer from 'react-test-renderer';
import Textarea from '../../src/components/Textarea';

describe('Textarea', () => {
  it('should render Textarea', () => {
    const tree = renderer.create(
      <Textarea
        size="128"
        color="primary"
        view="flat"
        disabled
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render Textarea with height', () => {
    const tree = renderer.create(
      <Textarea
        size="128"
        color="primary"
        view="flat"
        height="200"
        disabled
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
