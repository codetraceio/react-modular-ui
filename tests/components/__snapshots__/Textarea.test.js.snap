// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should render Textarea 1`] = `
<div
  className="c-textarea -size-128 -view-flat -color-primary -disabled"
>
  <textarea
    disabled={true}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    placeholder={undefined}
    style={Object {}}
    value={undefined}
  />
</div>
`;

exports[`Textarea should render Textarea with height 1`] = `
<div
  className="c-textarea -size-128 -view-flat -color-primary -disabled"
>
  <textarea
    disabled={true}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    placeholder={undefined}
    style={
      Object {
        "height": "200px",
      }
    }
    value={undefined}
  />
</div>
`;
